File: internal/redact/redact.go
// Package redact masks credentials embedded in git argument vectors before they
// are written to the forensic audit log.
//
// The backend still executes the real, unmodified arguments — only the STORED
// copy is masked — so a password baked into a clone URL or an Authorization
// header does not persist in the audit DB. This is a deliberate, narrow
// exception to gitc's raw-logging default: it strips only unambiguous secrets
// (URL userinfo, Authorization tokens) while preserving everything forensically
// useful (the command, host, path, and remaining flags). Broader redaction of
// the raw record stays out of band by design (see docs/BACKLOG.md).
package redact

import (
	"regexp"
	"strings"
)

const mask = "***"

// urlCred matches a URL scheme followed by userinfo and an '@', e.g.
// "https://user:pass@host" or "https://ghp_token@host".
var urlCred = regexp.MustCompile(`([a-zA-Z][a-zA-Z0-9+.\-]*://)([^/@\s]+)@`)

// authHeader matches an Authorization header value, e.g. "Authorization: Bearer x".
var authHeader = regexp.MustCompile(`(?i)(authorization:\s*[a-z]+\s+)(\S+)`)

// Args returns a copy of args with embedded credentials masked. The input slice
// is not modified.
func Args(args []string) []string {
	out := make([]string, len(args))
	for i, a := range args {
		out[i] = String(a)
	}

	return out
}

// String masks credentials found anywhere in s.
func String(s string) string {
	s = urlCred.ReplaceAllStringFunc(s, func(m string) string {
		sub := urlCred.FindStringSubmatch(m)
		scheme, userinfo := sub[1], sub[2]

		// user:pass -> keep the username, mask the secret; a single field
		// (typically a token used as the username) is masked whole.
		if i := strings.IndexByte(userinfo, ':'); i >= 0 {
			return scheme + userinfo[:i] + ":" + mask + "@"
		}

		return scheme + mask + "@"
	})

	return authHeader.ReplaceAllString(s, "${1}"+mask)
}

File: src/redact.rs
//! Port of Go `internal/redact` — mask credentials embedded in git argument
//! vectors before they are written to the forensic audit log.
//!
//! The backend still executes the real, unmodified arguments — only the STORED
//! copy is masked. It strips only unambiguous secrets (URL userinfo, Authorization
//! tokens) while preserving everything forensically useful (command, host, path,
//! remaining flags).

use regex::Regex;
use std::sync::OnceLock;

const MASK: &str = "***";

/// URL scheme + userinfo + `@`, e.g. `https://user:pass@host` / `https://tok@host`.
fn url_cred() -> &'static Regex {
    static RE: OnceLock<Regex> = OnceLock::new();
    RE.get_or_init(|| Regex::new(r"([a-zA-Z][a-zA-Z0-9+.\-]*://)([^/@\s]+)@").unwrap())
}

/// An Authorization header value, e.g. `Authorization: Bearer x`.
fn auth_header() -> &'static Regex {
    static RE: OnceLock<Regex> = OnceLock::new();
    RE.get_or_init(|| Regex::new(r"(?i)(authorization:\s*[a-z]+\s+)(\S+)").unwrap())
}

/// Go `Args` — a copy of `args` with embedded credentials masked; the input is
/// not modified.
pub fn args(args: &[String]) -> Vec<String> {
    args.iter().map(|a| string(a)).collect()
}

/// Go `String` — mask credentials found anywhere in `s`.
pub fn string(s: &str) -> String {
    let stripped = url_cred().replace_all(s, |caps: &regex::Captures| {
        let scheme = &caps[1];
        let userinfo = &caps[2];
        // user:pass -> keep the username, mask the secret; a single field
        // (typically a token used as the username) is masked whole.
        match userinfo.find(':') {
            Some(i) => format!("{scheme}{}:{MASK}@", &userinfo[..i]),
            None => format!("{scheme}{MASK}@"),
        }
    });
    auth_header()
        .replace_all(&stripped, format!("${{1}}{MASK}").as_str())
        .into_owned()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn string_masks_credentials() {
        let cases = [
            (
                "https://alice:s3cr3t@github.com/x.git",
                "https://alice:***@github.com/x.git",
            ),
            (
                "https://ghp_ABC123def456@github.com/x.git",
                "https://***@github.com/x.git",
            ),
            ("ssh://git:key@host:22/repo", "ssh://git:***@host:22/repo"),
            (
                "https://github.com/inovacc/gitc.git",
                "https://github.com/inovacc/gitc.git",
            ),
            ("status", "status"),
            (
                "Authorization: Bearer abcdef123",
                "Authorization: Bearer ***",
            ),
            (
                "authorization: Basic dXNlcjpwYXNz",
                "authorization: Basic ***",
            ),
        ];
        for (input, want) in cases {
            assert_eq!(string(input), want, "String({input:?})");
        }
    }

    #[test]
    fn args_masks_and_copies() {
        let input = vec![
            "clone".to_string(),
            "https://x:tok@github.com/r.git".to_string(),
        ];
        let got = args(&input);
        assert_eq!(got[1], "https://x:***@github.com/r.git");
        assert!(!got.join(" ").contains("tok@"), "token leaked: {got:?}");
        // must not mutate the input.
        assert_eq!(input[1], "https://x:tok@github.com/r.git");
    }
}
